### ISSUE TITLE
Regent: CUDA compiler breaks when passing large structures as an argument

### ISSUE THREAD

**Issue by mariodirenzo** [StanfordLegion/legion#958]
I'm encountering the following error 
```
...cfd/direnzo/legion/language/terra.build/src/terralib.lua:2581: too many results to unpack
stack traceback:
	[C]: in function 'unpack'
	...cfd/direnzo/legion/language/terra.build/src/terralib.lua:2581: in function 'docheck'
	...cfd/direnzo/legion/language/terra.build/src/terralib.lua:3039: in function 'checkexp'
	...cfd/direnzo/legion/language/terra.build/src/terralib.lua:2566: in function 'checkexpressions'
	...cfd/direnzo/legion/language/terra.build/src/terralib.lua:3235: in function 'checksingle'
	...cfd/direnzo/legion/language/terra.build/src/terralib.lua:3263: in function 'checkstmts'
	...cfd/direnzo/legion/language/terra.build/src/terralib.lua:3148: in function 'checkblock'
	...cfd/direnzo/legion/language/terra.build/src/terralib.lua:3347: in function 'typecheck'
	...cfd/direnzo/legion/language/terra.build/src/terralib.lua:1147: in function 'flattenkernel'
	.../cfd/direnzo/legion/language/terra.build/src/cudalib.lua:43: in function 'addkernel'
	.../cfd/direnzo/legion/language/terra.build/src/cudalib.lua:57: in function 'toptx'
	...atch/cfd/direnzo/legion/language/src/regent/cudahelper.t:309: in function 'jit_compile_kernels_and_register'
	/scratch/cfd/direnzo/legion/language/src/regent/std.t:4203: in function 'setup'
	/scratch/cfd/direnzo/legion/language/src/regent/std.t:4621: in function 'saveobj'
	prometeo-desugared.rg:1932: in main chunk
```
when I try to compile a task that takes a large structure (a few KB) as a constant input parameter.

**Comment by elliottslaughter** [comment]
I think this is an issue of a previously reported problem: https://github.com/terralang/terra/issues/152#issuecomment-452045621

**Comment by mariodirenzo** [comment]
Is there an existing workaround for this issue?

**Comment by elliottslaughter** [comment]
I posted a workaround here: https://github.com/terralang/terra/issues/152#issuecomment-717467429

**Comment by elliottslaughter** [comment]
@mariodirenzo Do you happen to have a reproducer for this issue?

I was thinking about this, particularly in relation to #694, but it just occurred to me that you said you're using structs up above and I want to confirm how you were even making the constant structs in the first place.

Thanks.

**Comment by mariodirenzo** [comment]
I can provide an older commit of the HTR solver that was triggering the issue.
In case you want, I can work on commenting out stuff so that it is faster at reproducing the issue.

**Comment by elliottslaughter** [comment]
If you can link me to the old version of the code that had the problem (and ideally the specific part of the code that had the problem), that may be enough for me to isolate the problem. Thanks.

**Comment by elliottslaughter** [comment]
I have confirmed that the issue no longer reproduces in `master` (b12759201c546e034877e59e448886c24207d7d3). It does reproduce in 02e79edf1b355dbc68ebf717cc2f88cb1cc4b3c0 with no further changes. I currently suspect this got fixed when we applied the solution for #638, but it might be good to check that this is true.

**Comment by elliottslaughter** [comment]
I have additionally confirmed that the error reproduces in 519127319f4efb20c7fcb35324ba27a1788deee2 but not in 72998c859640dcc159bb5a0498c9aaed4947d044 (i.e., the commits immediately before and after the fix to #638). I think this settles it pretty definitively; the problem must have been related to the formal parameter passing in Terra and as a result of using a different method to do that, is now gone.

**Comment by elliottslaughter** [comment]
For posterity I'm also attaching the steps for reproducing the error, though I don't currently expect to need them.

  * Use HTR private repo commit d66750a64aa7461c7d3c71119804ec381a7c5f98
  * Legion commit 519127319f4efb20c7fcb35324ba27a1788deee2 is the oldest that reproduces
  * Build with instructions from the HTR repo:
      * It's ok to use `install.py` instead of `setup_env.py`
      * LLVM does not matter (I used 11)
      * `USE_CUDA=1`
      * HDF and OpenMP do not matter (I had them off)
      * The exact Terra commit should not matter, I was using 0f54b3f916e6481eb72a54d6eef1e0c95c5fa65a (master at time of writing)
  * To quickly reproduce the issue it is sufficient to run `make prometeo_main_CH4_30SpMix.o`
